packages/amplify-cli-core/src/types.ts

```diff
@@ -20,19 +20,209 @@ export type $TSContext = {
+  emitError: (error: Error | null) => Promise<void>;
+  emitSuccess: () => Promise<void>;
+  init: (
+    installationUuid: string,
+    version: string,
+    input: $CommandLineInput,
+    accountId: string,
+    projectSettings: ProjectSettings,
+    processStartTimeStamp: number,
+  ) => void;
+  getUsageDataPayload: (error: Error | null, state: string) => IUsageDataPayload;
+  startCodePathTimer: (codePath: StartableTimedCodePath) => void;
+  stopCodePathTimer: (codePath: StoppableTimedCodePath) => void;
+  calculatePushNormalizationFactor: (events: { StackId: string; PhysicalResourceId: string }[], StackId: string) => void;
+  getSessionUuid: () => string;
+}
+
+export type StartableTimedCodePath = ManuallyTimedCodePath | UntilExitTimedCodePath;
+export type StoppableTimedCodePath = ManuallyTimedCodePath | FromStartupTimedCodePaths;
+
+export enum FromStartupTimedCodePaths {
+  PLATFORM_STARTUP = 'platformStartup', // time from CLI process start to plugin invoke. This timer is auto started on process startup
+  TOTAL_DURATION = 'totalDuration', // time from CLI process start to exit
+}
+export enum UntilExitTimedCodePath {
+  POST_PROCESS = 'postProcess', // time from plugin exit to process exit. This timer is automatically stopped at the end of the process
+}
+export enum ManuallyTimedCodePath {
+  // general paths (applies to all commands)
+  PLUGIN_TIME = 'pluginTime', // time spent in command handler plugin
+
+  // push-specific paths
+  PUSH_TRANSFORM = 'pushTransform', // total time spent transforming resources and uploading assets to prepare for a push
+  PUSH_DEPLOYMENT = 'pushDeployment', // time spent deploying CFN
+
+  // init-specific paths (also called during env checkout and pull)
+  INIT_ENV_PLATFORM = 'initEnvPlatform', // time to call awscloudformation provider initEnv. This includes downloading deployment bucket and updating local files
+  INIT_ENV_CATEGORIES = 'initEnvCategories', // time to call all of the category's initEnv methods
+
+  PROMPT_TIME = 'promptTime', // total time to takes to answer a prompt
+}
+
+export interface ContextParameters extends Pick<$CommandLineInput, 'argv' | 'plugin' | 'command' | 'options'> {
+  raw: $CommandLineInput['argv'];
+  array: $CommandLineInput['subCommands'];
+  first?: string;
+  second?: string;
+  third?: string;
+}
+
+export type $CommandLineInput = {
+  argv: Array<string>;
+  plugin?: string;
+  command: string;
+  subCommands?: string[];
+  options?: {
+    restore?: boolean;
+    json?: boolean;
+    name?: string;
+    awsInfo?: string;
+    config?: string;
+    'iterative-rollback'?: boolean;
+    force?: boolean;
+    env?: string;
+    rootStackName?: string;
+    frontend?: string;
+    quickstart?: boolean;
+    app?: string | boolean;
+    timeout?: string;
+    event?: string;
+    minify?: boolean;
+    help?: boolean;
+    localEnvFilePath?: string;
+    yes?: boolean;
+    appId?: string;
+  } & Record<string, string | boolean>;
```

Comment:
do all explicit options apply to every command ?

if not then perhaps we need `CliCommand` type and subtypes with coherent set of options?

Or we type options just as `Record<string, string | boolean>` and we let plugins to cast it down into something more specific.

Reply:
It gets pretty convoluted. In some cases, plugins are overwriting the options with values that go against their type. See https://github.com/sdstolworthy/amplify-cli/blob/0f94a470070a5e4ff6f2b4a1212f9f6a81e0fbd6/packages/amplify-cli/src/init-steps/postInitSetup.ts#L19.

Ultimately, I think the right course is to define options on `CommandLineInput` simply as a `Record<string, string (or unknown?)>`. Then, allow the plugins to downcast to a more specific type. Ultimately, `options` is just a collection of flags that the `cli` parses as a convenience for itself and its plugins. (see: https://github.com/sdstolworthy/amplify-cli/blob/0f94a470070a5e4ff6f2b4a1212f9f6a81e0fbd6/packages/amplify-cli/src/input-manager.ts#L40)

I think this defining the used flags here is a first step for now, but I could be talked out of that.

Reply:
I'd rather follow the downcast route and create more concrete types and mapping in plugins.

Reply:
Another vote in favor of defining this type as `Record<string, string | boolean>`

Reply:
That's still elevating command details to generic scope right ?. I.e. IDE would show all options in every plugin if I write `options.`.

Is it possible to add `getOptionsAs<T>()` instead of exposing property? (In Java that would be possible I recall).
So that plugins could use it to downcast to their specific type ?

If that's too hard to do in one go I'd recommend to:
- keep options and CLIOptionFlags but mark them deprecated (to not spread)
- expose new accessor/method/property that allows downcasting and make it work same way as `options`

Reply:
or maybe just deprecating that `CLIOptionFlags` right away is good enough.

Reply:
I'm still noodling on this. I don't mind the idea of deprecating the CLIOptionsFlags off the bat. But, like you said, I hesitate to bring them into the globally scoped package in the first place. A few of the flags are treated as global flags (yes, headless, frontend, version). [See here](https://github.com/sdstolworthy/amplify-cli/blob/tscontext-typing/packages/amplify-cli/src/input-manager.ts). 

I'm trying to think of a good way to do this that doesn't completely derail this PR.

Reply:
truely global flags might be defined in core and then plugins extend that type.

Reply:
Here's my current thinking: A union type between `CLIGlobalFlags & Record<string, any>`. This allows us to pull out the global flags, but not pollute core with an interface. Then, I can punt to a future PR a better implementation of a method that grabs a value from the options record and makes assertions about its type.

I think any kind of method to allow the plugins to assert types around the flags they expect needs to be given its own consideration in a separate request.

Reply:
If `CLIGlobalFlags` contains only global flags then I think this is good .